STM32Cube/Tasks/controller.c

```diff
@@ -0,0 +1,67 @@
+/*
+ * controller.c
+ *
+ *  Created on: Apr 14, 2024
+ *      Author: joedo
+ */
+
+#include "controller.h"
+#include "freertos.h"
+#include "cmsis_os.h" //TODO replace with xtaskDelayUntil
+#include "stm32h7xx_hal.h"
+#include <stdio.h>
+#include <math.h>
+#include "millis.h"
+
+extern UART_HandleTypeDef huart4;
+
+
+//needed dependecy for stdio-->printf()
+int __io_putchar(int ch)
+{
+	HAL_UART_Transmit(&huart4,(uint8_t*)&ch,1,HAL_MAX_DELAY);
+	return ch;
+}
+
+
+
+void controlTask(void *argument)
+{
+	float altitudeEstimate;
+	controller_t airbrakesController;
+	TickType_t last_ticks = xTaskGetTickCount();
+
+	float startTime = millis();
+	osDelay(1000);
+	float measuredDelay = millis()-startTime;
+	printf("1000ms HAL_Delay is: %lu ms\r\n",(unsigned long)(round(measuredDelay)));
+
+  /* Infinite loop */
+	for(;;)
+	{
+		//TODO get altitude message from queue
+		altitudeEstimate = 1000; //Test code
+
+		//PID controller update
+		airbrakesController.error = airbrakesController.target_altitude - altitudeEstimate;
+		airbrakesController.controller_term_P = airbrakesController.error * CONTROLLER_GAIN_P;
+		float dt = (float)(xTaskGetTickCount() - last_ticks) / (portTICK_RATE_MS * 1000); //time delay in ms TODO: convert to more accurate source
```

Comment:
can you replace this with millis()? wasn't that sorta the point?

Reply:
That was indeed the point, fixed.